### PR TITLE
Reusable UI components

### PR DESCRIPTION
**Pull request by snixtho**
Adds:
- Window
- Alert
- Button
- Checkbox
- Dropdown
- IconButton
- RadioButton
- Switch
- TextInput
- HiddenEntry
- Common UI scripts
- Event-based UI

Also moved the MOTD onto the new window system. There is also a UI demo when a user types the /uidemo command (be in fullscreen otherwise you may see glitches and defects).

### CODE REVIEW

**Review by Chris92de — APPROVED** [review]

**Review by AtomicLiquid — APPROVED** [review]

File: src/EvoSC.Manialinks/Templates/Controls/Alert.mt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    
+    <!-- The text to display in the alert. -->
+    <property type="string" name="text" default="This is an alert" />
+    
+    <template>
+        <frame 
+                pos="{{ x }} {{ y }}" 
+                size="{{ width+8 }} 7" 
+                id="{{ id }}" 
+                data-width="{{ width+7 }}"
```
Comment (brakerb) [inline src/EvoSC.Manialinks/Templates/Controls/Alert.mt:28]: Is there a reason why it is `+7` here and `+8` above?

File: src/EvoSC.Manialinks/Templates/Scripts/Alert.ms [hunk 2]
```diff
@@ -0,0 +1,44 @@
+﻿<!--
+Void ShowAlert(Text alertId) {
+    declare frame <=> Page.MainFrame.GetFirstChild(alertId) as CMlFrame;
+    declare frameInner <=> Page.MainFrame.GetFirstChild(alertId^"-inner-frame") as CMlFrame;
+    declare width = TextLib::ToInteger(frame.DataAttributeGet("width"));
```
Comment (brakerb) [inline src/EvoSC.Manialinks/Templates/Scripts/Alert.ms:5]: Couldn't we use `frame.Size.X - 1`?

File: src/EvoSC.Manialinks/Templates/Scripts/Checkbox.ms [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    declare bg <=> checkboxFrame.Controls[4] as CMlQuad;
+    declare icon <=> checkboxFrame.Controls[6] as CMlLabel;
+    
+    IsChecked = value;
+    
+    if (value) {
+        AnimMgr.Add(bg, """<quad opacity='1' />""", 500, CAnimManager::EAnimManagerEasing::ExpOut);
+        icon.SetText("");
+    } else {
+        AnimMgr.Add(bg, """<quad opacity='0' />""", 500, CAnimManager::EAnimManagerEasing::ExpOut);
```
Comment (brakerb) [inline src/EvoSC.Manialinks/Templates/Scripts/Checkbox.ms:15]: I found 320ms to be the sweetspot for transitions like these, but that's just a opinion

File: src/EvoSC.Manialinks/Templates/Scripts/Window.ms [hunk 4]
```diff
@@ -0,0 +1,109 @@
+﻿<!--
+Void CloseWindow(Text windowId) {
+    declare Window = Page.MainFrame.GetFirstChild(windowId) as CMlFrame;
+    Window.Hide();
+    
+    +++ OnWindowClosed +++
```
Comment (brakerb) [inline src/EvoSC.Manialinks/Templates/Scripts/Window.ms:6]: We should introduce a global script variable to trigger exiting the script. Even if the main-frame is hidden, the script will continue to run doing yields forever, unless we call a `return;` in the main-loop.

**Review by brakerb — APPROVED** [review]
Looks good overall, just a few concerns/suggestions

File: src/EvoSC.Manialinks/Templates/Controls/Alert.mt (hunk 1 above)
Comment (snixtho) [reply src/EvoSC.Manialinks/Templates/Controls/Alert.mt:28]: It would clip the entire thing by 1 unit if not for some reason.

**Pull request merged**